common/switchboard.hpp

```diff
@@ -1,179 +1,571 @@
-*/
+const record_header __switchboard_topic_stop_header {"switchboard_topic_stop", {
+    {"plugin_id", typeid(plugin_id_t)},
+    {"topic_name", typeid(std::string)},
+    {"enqueued", typeid(std::size_t)},
+    {"dequeued", typeid(std::size_t)},
+    {"idle_cycles", typeid(std::size_t)},
+}};
 
 /**
- * @brief A manager for typesafe, threadsafe, named event-streams (called topics).
+ * @brief A manager for typesafe, threadsafe, named event-streams (called
+ * topics).
  *
- * - Writing: One can write to a topic (in any thread) through the `ILLIXR::writer` returned by
- *   `publish()`.
- * 
- * - There are two ways of reading: asynchronous reading and synchronous reading:
+ * - Writing: One can write to a topic (in any thread) through the
+ * object returned by `get_writer()`.
  *
- *   - Asynchronous reading returns the most-recent event on the topic (idempotently). One can do
- *     this through (in any thread) the `ILLIXR::reader_latest` handle returned by
- *     `subscribe_latest()`.
+ * - There are two ways of reading: asynchronous reading and synchronous
+ * reading:
  *
- *   - Synchronous reading schedules a callback to be executed on _every_ event which gets
- *     published. One can schedule computation by `schedule()`, which will run the computation in a
- *     thread managed by switchboard.
+ *   - Asynchronous reading returns the most-recent event on the topic
+ * (idempotently). One can do this through (in any thread) the
+ * object returned by `get_reader()`.
+ *
+ *   - Synchronous reading schedules a callback to be executed on _every_ event
+ * which gets published. One can schedule computation by `schedule()`, which
+ * will run the computation in a thread managed by switchboard.
  *
  * \code{.cpp}
- * void do_stuff(switchboard* sb) {
- *     auto topic1 = sb->subscribe_latest<topic1_type>("topic1");
- *     auto topic2 = sb->publish<topic2_type>("topic2");
- * 
- *     // Read topic 3 synchronously
- *     sb->schedule<topic3_type>("topic3", [&](const topic3_type *event3) {
- *         // This is a lambda expression
- *         // https://en.cppreference.com/w/cpp/language/lambda
- *         std::cout << "Got a new event on topic3: " << event3 << std::endl;
- *     });
+ * // Get a reader on topic1
+ * switchboard::reader<topic1_type> topic1 = switchboard.get_reader<topic1_type>("topic1");
  *
- *     while (true) {
- *         // Read topic 1
- *         topic1_type* event1 = topic1.get_latest_ro();
+ * // Get a writer on topic2
+ * switchboard::writer<topic2_type> topic2 = switchboard.get_reader<topic2_type>("topic1");
  *
- *         // Write to topic 2
- *         topic2_type* event2 = new topic2_type;
- *         topic2.put(event2);
- *     }
+ * while (true) {
+ *     // Read topic 1
+ *     switchboard::ptr<topic1_type> event1 = topic1.get();
+ *
+ *     // Write to topic 2
+ *     topic2_type* event2 = topic2.allocate();
+ *     // Populate the event
+ *     event2->foo = 3;
+ *     topic2.put(event2);
```

Comment:
Oh, that's because in practice we also use `std::shared_ptr`. We should switch this example to that. (`std::move`ing a raw pointer doesn't make sense).